File: typescript/base/src/main/vsdk/toolkit/environment/geometry/elements/PatchSet.ts
import { Ray } from "../../../common/linealAlgebra/Ray";
import { Geometry } from "../../../skin/Geometry";
import { GeometryClassId } from "../../../skin/GeometryClassId";
import type { Patch } from "./Patch";
import type { RayHit } from "./RayHit";

export class PatchSet extends Geometry {
  private patchList: Patch[] | null;
  private memoryPoolManaged: boolean;

  public constructor(input: Patch[] | null) {
    super(GeometryClassId.PATCH_SET);
    this.memoryPoolManaged = false;
    this.patchList = [];
    for (let i = 0; input !== null && i < input.length; i++) {
      const patch = input[i];
      if (patch !== undefined) {
        this.patchList.push(patch);
      }
    }

    Geometry.patchListBounds(this.getPatchList(), this.boundingBox);
    this.boundingBox.enlargeTinyBit();
    this.bounded = true;
  }

  public override destroy(): void {
    if (this.patchList !== null) {
      this.patchList = null;
    }
    super.destroy();
  }

  public override discretizationIntersect(
    ray: Ray,
    minimumDistance: number,
    maximumDistance: number[],
    hitFlags: number,
    hitStore: RayHit | null
  ): RayHit | null {
    if (!this.discretizationIntersectPreTest(ray, minimumDistance, maximumDistance)) {
      return null;
    }

    return Geometry.patchListIntersect(this.patchList, ray, minimumDistance, maximumDistance, hitFlags, hitStore);
  }

  public getPatchList(): Patch[] | null {
    return this.patchList;
  }

  public isMemoryPoolManaged(): boolean {
    return this.memoryPoolManaged;
  }

  public setMemoryPoolManaged(value: boolean): void {
    this.memoryPoolManaged = value;
  }
}

File: cpp/base/src/main/vsdk/toolkit/environment/geometry/elements/PatchSet.cpp
#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/environment/geometry/elements/PatchSet.h"

PatchSet::PatchSet(const java::ArrayList<Patch *> *input): Geometry(GeometryClassId::PATCH_SET) {
    memoryPoolManaged = false;
    patchList = new java::ArrayList<Patch *>();
    for ( int i = 0; input != nullptr && i < input->size(); i++ ) {
        patchList->add(input->get(i));
    }

    Geometry::patchListBounds(getPatchList(), &boundingBox);
    boundingBox.enlargeTinyBit();
    bounded = true;
}

PatchSet::~PatchSet() {
    // PatchSet always owns its internal patch pointer container.
    // `isDuplicate` only controls shared resources at Geometry level
    // (for example radianceData), not this list's lifetime.
    if ( patchList != nullptr ) {
        delete patchList;
        patchList = nullptr;
    }
}

/**
DiscretizationIntersect returns nullptr is the ray doesn't hit the discretization
of the object. If the ray hits the object, a hit record is returned containing
information about the intersection point. See geometry.h for more explanation

Tests whether the Ray intersect the patches in the list. See geometry.h
(GeomDiscretizationIntersect()) for more explanation
*/
RayHit *
PatchSet::discretizationIntersect(
    Ray *ray,
    float minimumDistance,
    float *maximumDistance,
    int hitFlags,
    RayHit *hitStore) const
{
    if ( !discretizationIntersectPreTest(ray, minimumDistance, maximumDistance) ) {
        return nullptr;
    }

    return Geometry::patchListIntersect(patchList, ray, minimumDistance, maximumDistance, hitFlags, hitStore);
}

java::ArrayList<Patch *> *
PatchSet::getPatchList() const {
    return patchList;
}

bool
PatchSet::isMemoryPoolManaged() const {
    return memoryPoolManaged;
}

void
PatchSet::setMemoryPoolManaged(const bool value) {
    memoryPoolManaged = value;
}
